lunamark/reader/markdown.lua

```diff
@@ -542,18 +544,23 @@ parsers.urlchar = parsers.anyescaped - parsers.newline - parsers.more
 
 parsers.Block        = V("Block")
 
+-- For fenced divs, the closing fence must match the opening fence, since
+-- divs of increasing fence length can be nested.
+-- But for fenced code blocks, "the closing fence must be at least as long as
+-- the opening fence".
+
 parsers.TildeFencedCodeBlock
-                     = parsers.fencehead(parsers.tilde)
+                     = parsers.fencehead(parsers.tilde, captures_geq_length)
                      * Cs(parsers.fencedline(parsers.tilde)^0)
                      * parsers.fencetail(parsers.tilde)
 
 parsers.BacktickFencedCodeBlock
-                     = parsers.fencehead(parsers.backtick)
+                     = parsers.fencehead(parsers.backtick, captures_geq_length)
                      * Cs(parsers.fencedline(parsers.backtick)^0)
                      * parsers.fencetail(parsers.backtick)
 
 parsers.ColonFencedDivBlock
-                     = parsers.fencehead(parsers.colon)
+                     = parsers.fencehead(parsers.colon, captures_equal_length)
```

Comment:
Is this right? Pandoc is quite happy to parse the following:

``` md
::: foo
bar
:::::::
```

Reply:
Here is the corresponding code in Pandoc's Haskell code:

https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Readers/Markdown.hs#L2066-L2089

I am not fluent in Haskell, but here is what I could glean:

- Pandoc does not care about the number of `:::` we use:

    ``` haskell
      string ":::"
      skipMany (char ':')
    ```

- Pandoc uses an entirely different logic for parsing fenced code and fenced divs. With fenced code, there is no internal structure, so counting backticks/tildes is the best we can hope for. With fenced divs, the inside of the div is Markdown text, so we can detect nesting without counting colons. In the Haskell code, this is tracked by the `stateFencedDivLevel` variable. This variable is then interspersed throughout the grammar.